File: backend/app/schemas/practice.py
from datetime import datetime
from pydantic import BaseModel, ConfigDict, Field

class PracticeSubmission(BaseModel):
    model_config = ConfigDict(str_strip_whitespace=True)
    answer_text: str = Field(min_length=20, max_length=12000)
    follow_up_from: int | None = None

class PracticeDraft(BaseModel):
    draft_text: str = Field(default='', max_length=12000)
    bookmarked: bool = False
    follow_up_from: int | None = None

class Feedback(BaseModel):
    relevance: int = Field(ge=1, le=5)
    specificity: int = Field(ge=1, le=5)
    structure: int = Field(ge=1, le=5)
    technical_depth: int = Field(ge=1, le=5)
    strengths: list[str] = Field(min_length=1, max_length=4)
    improvements: list[str] = Field(min_length=1, max_length=4)
    next_step: str = Field(min_length=1, max_length=1500)
    follow_up: str = Field(min_length=1, max_length=1000)

class PracticeAttemptResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: int
    question_id: int
    answer_text: str
    prompt_text: str
    feedback_json: Feedback
    created_at: datetime

File: frontend/src/api/practice.ts
import {getAuthHeaders, handleResponse} from './auth';
export type Feedback = {relevance: number; specificity: number; structure: number; technical_depth: number; strengths: string[]; improvements: string[]; next_step: string; follow_up: string};
export type Attempt = {id: number; question_id: number; answer_text: string; prompt_text: string; feedback_json: Feedback; created_at: string};
const API_URL = import.meta.env.VITE_API_URL;
export async function fetchAttempts(id: number): Promise<Attempt[]> {
  return handleResponse(await fetch(`${API_URL}/answers/${id}/attempts`, {headers: getAuthHeaders()}));
}
export async function submitAttempt(id: number, answer_text: string, follow_up_from?: number): Promise<Attempt> {
  return handleResponse(await fetch(`${API_URL}/answers/${id}/attempts`, {method: 'POST', headers: {...getAuthHeaders(), 'Content-Type': 'application/json'}, body: JSON.stringify({answer_text, follow_up_from})}));
}
export async function saveDraft(id: number, draft_text: string, bookmarked: boolean) {
  return handleResponse(await fetch(`${API_URL}/answers/${id}/practice`, {method: 'PATCH', headers: {...getAuthHeaders(), 'Content-Type': 'application/json'}, body: JSON.stringify({draft_text, bookmarked})}));
}
